mrbgems/picoruby-net/mrblib/net.rb

```diff
@@ -0,0 +1,155 @@
+              headers[fields[0]] = fields[1]&.strip
+            end
+          else
+            body += lines[line_number]
+            body += "\r\n" unless lines[line_number + 1].nil?
+          end
+          line_number += 1
+        end
+        return {
+          status: status,
+          headers: headers,
+          body: body
+        }
+      else
+        return nil
+      end
+    end
+  end
+
+  class DNS
+  end
+
+  class TCPClient
+    def self.request(host, port, content, is_tls)
+      # ip = DNS.resolve(host)
+      _request_impl(host, port, content, is_tls)
+    end
+  end
+
+  class HTTPClient
+    def initialize(host)
+      @host = host
+    end
+
+    def get(path)
+      req =  "GET #{path} HTTP/1.1\r\n"
+      req += "Host:#{@host}\r\n"
+      req += "Connection: close\r\n"
+      req += "\r\n"
+
+      return Net::HTTPUtil.format_response(TCPClient.request(@host, 80, req, false))
+    end
+
+    def get_with_headers(path, headers)
+      req =  "GET #{path} HTTP/1.1\r\n"
+      req += "Host:#{@host}\r\n"
+      req += "Connection: close\r\n"
+      headers.each do |k, v|
+        req += "#{k}: #{v}"
+      end
+      req += "\r\n"
+
+      return Net::HTTPUtil.format_response(TCPClient.request(@host, 80, req, false))
+    end
+
+    def post(path, headers, body)
+      req =  "POST #{path} HTTP/1.1\r\n"
+      req += "Host:#{@host}\r\n"
+      req += "Connection: close\r\n"
+      headers.each do |k, v|
+        req += "#{k}: #{v}"
+      end
+      req += "#{body}\r\n"
+      req += "\r\n"
+
+      return Net::HTTPUtil.format_response(TCPClient.request(@host, 80, req, false))
+    end
+
+    def put(path, headers, body)
+      req =  "PUT #{path} HTTP/1.1\r\n"
+      req += "Host:#{@host}\r\n"
+      req += "Connection: close\r\n"
+      headers.each do |k, v|
+        req += "#{k}: #{v}"
+      end
+      req += "#{body}\r\n"
+      req += "\r\n"
+
+      return Net::HTTPUtil.format_response(TCPClient.request(@host, 80, req, false))
+    end
+  end
+
+  class HTTPSClient
+    def initialize(host)
+      @host = host
+    end
+
+    def get(path)
+      req =  "GET #{path} HTTP/1.1\r\n"
+      req += "Host:#{@host}\r\n"
+      req += "Connection: close\r\n"
+      req += "\r\n"
+
+      return Net::HTTPUtil.format_response(TCPClient.request(@host, 443, req, true))
+    end
+
+    def get_with_headers(path, headers)
+      req =  "GET #{path} HTTP/1.1\r\n"
+      req += "Host:#{@host}\r\n"
+      req += "Connection: close\r\n"
+      headers.each do |k, v|
+        req += "#{k}: #{v}"
+      end
+      req += "\r\n" 
+
+      return Net::HTTPUtil.format_response(TCPClient.request(@host, 443, req, true))
+    end
+
+    def post(path, headers, body)
+      req =  "POST #{path} HTTP/1.1\r\n"
+      req += "Host:#{@host}\r\n"
+      req += "Connection: close\r\n"
+      headers.each do |k, v|
+        req += "#{k}: #{v}"
+      end
+      req += "#{body}\r\n"
+      req += "\r\n"
```

Comment:
```suggestion
      headers.each do |k, v|
        req += "#{k}: #{v}\r\n"
      end
      req += "\r\n"
      req += body
```

- I tested `Net::HTTPSClient#post` using this branch with an IoT PaaS: "ambidata.io", but it didn't work if the header contained `Connection: close`, so I'd like to make this optional
- A linebreak, `\r\n` is required for each header
- Another linebreak is required between the header and the body, I think
- Instead, if linebreaks are automatically inserted after the `#{body}`, the Content-Length calculation by the user will conflict, so it's better to let the user manage whether some linebreaks are required

Reply:
Thanks for the review! The patch looks good so I will apply it. These are the reasoning:

> A linebreak, \r\n is required for each header
> Another linebreak is required between the header and the body, I think

Yes, these are correct according to [RFC 9112 Section 2.1](https://datatracker.ietf.org/doc/html/rfc9112#name-message-format) and [Section 5](https://datatracker.ietf.org/doc/html/rfc9112#name-field-syntax). 

Also, [`Connection: close` is a SHOULD](https://datatracker.ietf.org/doc/html/rfc9112#name-tear-down) in HTTP/1.1, but [not allowed in HTTP/2](https://developer.mozilla.org/ja/docs/Web/HTTP/Headers/Connection). This might be the reason it didn't work with `ambidata.io` (they might be assuming all requests to be HTTP/2-compliant). As it is optional it might be better to make it optional in our implementation.

Reply:
- Applied changes to the rest in 918e780
- Minor styling fix but added a whitespace between `Host:` header's colon and content in 65f6d67